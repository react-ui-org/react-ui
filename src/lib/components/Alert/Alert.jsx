import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootColorClassName } from '../_helpers/getRootColorClassName';
import styles from './Alert.scss';

export const Alert = ({
  children,
  color,
  icon,
  id,
  onClose,
  translations,
}) => (
  <div
    className={classNames(
      styles.root,
      getRootColorClassName(color, styles),
    )}
    id={id}
    role="alert"
  >
    {icon && (
      <div className={styles.icon}>
        {icon}
      </div>
    )}
    <div
      className={styles.message}
      {...(id && { id: `${id}__content` })}
    >
      {children}
    </div>
    {onClose && (
      <button
        type="button"
        {...(id && { id: `${id}__close` })}
        className={styles.close}
        onClick={() => onClose()}
        onKeyPress={() => onClose()}
        tabIndex="0"
        title={translations.close}
      >
        <span className={styles.closeSign}>Ã—</span>
      </button>
    )}
  </div>
);

Alert.defaultProps = {
  color: 'note',
  icon: null,
  id: undefined,
  onClose: null,
};

Alert.propTypes = {
  /**
   * Alert body.
   */
  children: PropTypes.node.isRequired,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the alert.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Optional element to be displayed next to the alert body.
   */
  icon: PropTypes.node,
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__close`
   * * `<ID>__content`
   */
  id: PropTypes.string,
  /**
   * Function to call when the close button is clicked. If not provided, close buttons will be
   * hidden.
   */
  onClose: PropTypes.func,
  /**
   * Translations required by the component.
   */
  translations: PropTypes.shape({
    close: PropTypes.string.isRequired,
  }).isRequired,
};

export const AlertWithContext = withProviderContext(Alert, 'Alert');

export default AlertWithContext;
