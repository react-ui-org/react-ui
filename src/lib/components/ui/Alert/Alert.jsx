import PropTypes from 'prop-types';
import React from 'react';
import getRootColorClassName from '../../../helpers/getRootColorClassName';
import { withProviderContext } from '../../../provider';
import styles from './Alert.scss';

export const Alert = ({
  children,
  closeHandler,
  color,
  icon,
  id,
  translations,
}) => (
  <div
    className={[
      styles.root,
      getRootColorClassName(color, styles),
    ].join(' ')}
    id={id}
    role="alert"
  >
    {icon && (
      <div className={styles.icon}>
        {icon}
      </div>
    )}
    <div
      className={styles.message}
      {...(id && { id: `${id}__content` })}
    >
      {children}
    </div>
    {closeHandler && (
      <button
        type="button"
        {...(id && { id: `${id}__close` })}
        className={styles.close}
        onClick={() => closeHandler()}
        onKeyPress={() => closeHandler()}
        tabIndex="0"
        title={translations.close}
      >
        <span className={styles.closeSign}>Ã—</span>
      </button>
    )}
  </div>
);

Alert.defaultProps = {
  closeHandler: null,
  color: 'note',
  icon: null,
  id: undefined,
};

Alert.propTypes = {
  /**
   * Alert body.
   */
  children: PropTypes.node.isRequired,
  /**
   * Function to call when the close button is clicked. If not provided, close buttons will be
   * hidden.
   */
  closeHandler: PropTypes.func,
  /**
   * [Color variant](/foundation/colors#component-colors) to clarify importance and meaning of the alert.
   */
  color: PropTypes.oneOf(
    ['primary', 'secondary', 'success', 'warning', 'danger', 'help', 'info', 'note', 'light', 'dark'],
  ),
  /**
   * Optional element to be displayed next to the alert body.
   */
  icon: PropTypes.node,
  /**
   * ID of the root HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__close`
   * * `<ID>__content`
   */
  id: PropTypes.string,
  /**
   * Translations required by the component.
   */
  translations: PropTypes.shape({
    close: PropTypes.string.isRequired,
  }).isRequired,
};

export const AlertWithContext = withProviderContext(Alert, 'Alert');

export default AlertWithContext;
