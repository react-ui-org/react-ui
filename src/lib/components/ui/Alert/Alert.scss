@import '../../../styles/tools/reset';
@import '../../../styles/tools/spacing';
@import 'mixins';

.root {
  position: relative;
  display: flex;
  align-items: flex-start;
  width: 100%;
  border-width: $alert-border-width $alert-border-width $alert-border-width $alert-stripe-width;
  border-style: solid;
  border-radius: $alert-border-radius;
}

.icon,
.message {
  padding: $alert-padding;
}

.close,
.icon {
  height: $alert-min-height;
}

.icon {
  display: flex;
  flex: none;
  align-items: center;
  justify-content: center;
  padding-right: 0;
}

.message {
  flex-grow: 1;
  font-weight: $alert-font-weight;
  font-size: $alert-font-size;
  line-height: $alert-line-height;
}

.message a,
.message strong {
  font-weight: $alert-emphasis-font-weight;
}

.close {
  @include reset-button();

  padding: $alert-padding;
  font-size: map-get($typography-size-values, 3);
  line-height: 1;
}

.closeSign {
  position: relative;
  top: -0.1em;
}

.isRootSuccess {
  @include alert-type(success);
}

.isRootWarning {
  @include alert-type(warning);
}

.isRootError {
  @include alert-type(error);
}

.isRootInfo {
  @include alert-type(info);
}

.isRootHelp {
  @include alert-type(help);
}

.isRootNote {
  @include alert-type(note);
}
