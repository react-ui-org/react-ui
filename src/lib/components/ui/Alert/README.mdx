---
name: Alert
menu: 'Components: UI'
route: /components/ui/alert
---

# Alert

import {
  Playground,
  Props,
} from 'docz'
import Icon from '../../../../docs/_components/Icon'
import Button from '../Button'
import {
  AlertWithContext as Alert,
  Alert as ParsableAlert,
} from './Alert'

Alert presents feedback or important information to users.

## Basic Usage

To implement the Alert component, you need to import it first:

```js
import { Alert } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Alert>
    Hello!
  </Alert>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- **Be simple and short** and explain to users why you are interrupting them
  with an alert. Clearly say what happened or what is going to happen. Provide a
  brief title when the message is too long.

- **Use icons** to improve the accessibility of alerts because color may not be
  enough for everyone. [See how](#alerts-with-icons).

## Alert Types

There are six types of alerts (matching all available
[UI colors](/foundation/colors#ui-colors)) to cover all possible needs of your
project: note, success, warning, error, help, and info.

### Note (default)

When no type is specified, a neutral informative alert appears:

<Playground>
  <Alert>
    <strong>Note:</strong> This feature may not be available in all regions.
  </Alert>
</Playground>

### Success

Success alerts confirm that an operation has been processed successfully.

<Playground>
  <Alert type="success">
    <strong>Success:</strong> Settings have been successfully saved.
  </Alert>
</Playground>

### Warning

Use warning alerts when a potentially unfavourable situation may occur. You may
suggest an action to resolve the problem.

<Playground>
  <Alert type="warning">
    <strong>Warning:</strong> Your credit card is going to expire soon.
    {' '}
    <a href="/">Update my payment options</a>
  </Alert>
</Playground>

### Error

Error alerts say there is something that block users from continuing that
requires their immediate attention. The alert should offer a solution to the
problem.

<Playground>
  <Alert type="error">
    <strong>Error:</strong> Cannot connect to the server. Is your internet
    connection working fine?
    {' '}
    <a href="/">Try again</a>
  </Alert>
</Playground>

### Help

This kind of alert can be used to display helpful information.

<Playground>
  <Alert type="help">
    <strong>Help:</strong> You should choose a password you don&apos;t use
    anywhere else.
  </Alert>
</Playground>

### Info

Another common, informative alert.

<Playground>
  <Alert type="info">
    <strong>Info:</strong> This feature depends on user&apos;s OS preferences.
  </Alert>
</Playground>

## Alerts with Icons

An icon can (and should) accompany the message.

ðŸ‘‰ Please note there are no icons pre-packed in React UI. Visit
[Icons](/foundation/icons) to see how it works.

<Playground>
  <Alert type="success" icon={<Icon icon="success" />}>
    <strong>Success:</strong> Settings have been successfully saved.
  </Alert>
</Playground>

## Dismissible Alerts

You can make an alert dismissible by providing a function that closes it on
click on the close button:

<Playground>
  {() => {
    const [isAlertVisible, setIsAlertVisible] = React.useState(true);
    return (
      <>
        {isAlertVisible ? (
          <Alert
            type="success"
            icon={<Icon icon="success" />}
            closeHandler={() => setIsAlertVisible(false)}
          >
            <strong>Success:</strong> Settings have been successfully saved.
          </Alert>
        ) : (
          <Button
            clickHandler={() => setIsAlertVisible(true)}
            label="Show alert"
          />
        )}
      </>
    );
  }}
</Playground>

## API

<Props of={ParsableAlert} />

---

Next: [Badge â†’](/components/ui/badge)
