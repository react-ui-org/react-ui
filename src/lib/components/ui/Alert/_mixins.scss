@import 'variables';

@mixin alert-type($type) {
  $type-properties: map-get($alert-types, $type);
  $foreground-color: map-get($type-properties, foreground-color);

  color: map-get($type-properties, text-color);
  border-color: $foreground-color;
  background-color: map-get($type-properties, background-color);

  a,
  strong,
  .close,
  .icon {
    color: $foreground-color;
  }
}
