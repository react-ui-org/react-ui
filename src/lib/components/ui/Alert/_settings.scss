@use '../../../styles/settings/typography';
@use 'theme';

$font-size: map-get(typography.$size-values, 0);
$line-height: 1.5;
$min-height: calc(#{$font-size} * #{$line-height} + 2 * #{theme.$padding});
