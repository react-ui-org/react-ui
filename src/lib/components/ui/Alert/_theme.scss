$alert-border-width: var(--rui-alert-border-width);
$alert-border-radius: var(--rui-alert-border-radius);
$alert-padding: var(--rui-alert-padding);
$alert-icon-size: var(--rui-icon-size-default);

$alert-types: (
  success: (
    background-color: var(--rui-alert-success-background-color),
    foreground-color: var(--rui-alert-success-foreground-color),
    text-color: var(--rui-alert-success-text-color),
  ),
  warning: (
    background-color: var(--rui-alert-warning-background-color),
    foreground-color: var(--rui-alert-warning-foreground-color),
    text-color: var(--rui-alert-warning-text-color),
  ),
  error: (
    background-color: var(--rui-alert-error-background-color),
    foreground-color: var(--rui-alert-error-foreground-color),
    text-color: var(--rui-alert-error-text-color),
  ),
  info: (
    background-color: var(--rui-alert-info-background-color),
    foreground-color: var(--rui-alert-info-foreground-color),
    text-color: var(--rui-alert-info-text-color),
  ),
  help: (
    background-color: var(--rui-alert-help-background-color),
    foreground-color: var(--rui-alert-help-foreground-color),
    text-color: var(--rui-alert-help-text-color),
  ),
  note: (
    background-color: var(--rui-alert-note-background-color),
    foreground-color: var(--rui-alert-note-foreground-color),
    text-color: var(--rui-alert-note-text-color),
  ),
);
