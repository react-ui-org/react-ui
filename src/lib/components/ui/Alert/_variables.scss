@import '../../../styles/settings/typography';
@import './theme';

$alert-font-size: map-get($typography-size-values, 0);
$alert-line-height: 1.5;
$alert-min-height: calc(#{$alert-font-size} * #{$alert-line-height} + 2 * #{$alert-padding});
$alert-stripe-width: 5px;
$alert-message-font-weight: map-get($typography-font-weight-values, light);
$alert-title-font-weight: map-get($typography-font-weight-values, bold);
