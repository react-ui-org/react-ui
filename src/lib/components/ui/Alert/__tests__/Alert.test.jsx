import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Alert from '..';

describe('rendering', () => {
  it('renders correctly', () => {
    const tree = mount(
      <Alert>
        <div>Children</div>
      </Alert>,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = mount(
      <Alert
        closeHandler={() => {}}
        icon={<span className="icon" />}
        id="custom-id"
        type="success"
      >
        <div>Children</div>
      </Alert>,
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls closeHandler() on Close button click', () => {
    const spy = sinon.spy();
    const component = mount((
      <Alert
        closeHandler={spy}
        type="success"
      >
        <div>Children</div>
      </Alert>
    ));

    component.find('button').at(0).simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});
