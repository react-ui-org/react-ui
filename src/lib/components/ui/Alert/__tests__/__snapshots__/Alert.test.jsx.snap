// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<WithTranslationContextComponent
  translations={null}
>
  <Alert
    closeHandler={null}
    icon={null}
    translations={
      Object {
        "close": "Close",
      }
    }
    type="note"
  >
    <div
      className="root isRootNote"
      role="alert"
    >
      <div
        className="message"
      >
        <div>
          Children
        </div>
      </div>
    </div>
  </Alert>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with all props 1`] = `
<WithTranslationContextComponent
  closeHandler={[Function]}
  icon={
    <span
      className="icon"
    />
  }
  id="custom-id"
  translations={null}
  type="success"
>
  <Alert
    closeHandler={[Function]}
    icon={
      <span
        className="icon"
      />
    }
    id="custom-id"
    translations={
      Object {
        "close": "Close",
      }
    }
    type="success"
  >
    <div
      className="root isRootSuccess"
      id="custom-id"
      role="alert"
    >
      <div
        className="icon"
      >
        <span
          className="icon"
        />
      </div>
      <div
        className="message"
        id="custom-id__content"
      >
        <div>
          Children
        </div>
      </div>
      <button
        className="close"
        id="custom-id__close"
        onClick={[Function]}
        onKeyPress={[Function]}
        tabIndex="0"
        title="Close"
        type="button"
      >
        <span
          className="closeSign"
        >
          Ã—
        </span>
      </button>
    </div>
  </Alert>
</WithTranslationContextComponent>
`;
